### PR TITLE
DAO-202 Integrate policy downgrades

### PR DESCRIPTION
**Pull request by mcoetzee**
### What does this change?
- pulls in latest changes from https://github.com/api3dao/claims-manager/pull/65 
- adds a `downgrade-user-policy` Hardhat task
- includes the `DowngradedPolicy` log in the policy loading strategy

### How did you action this task?
We combine the `UpgradedPolicy` and `DowngradedPolicy` events, sort them, and use the last (most recent) event in order to get the latest `claimsAllowedUntil` value.

### CODE REVIEW

File: hardhat/hardhat.config.ts [hunk 1]
```diff
@@ -132,8 +131,8 @@ task('create-user-policy', 'Creates a policy for the given user')
       userAddress,
       userAddress,
       parseUsd(args.coverageAmount),
-      BigNumber.from(Math.round(claimsAllowedFrom.getTime() / 1000)),
-      BigNumber.from(Math.round(claimsAllowedUntil.getTime() / 1000)),
+      Math.round(claimsAllowedFrom.getTime() / 1000),
```
Comment (Siegrift) [inline hardhat/hardhat.config.ts:134]: I guess this was just to reduce characters, right?

File: src/logic/policies/data.ts [hunk 2]
```diff
@@ -124,27 +124,30 @@ export function useUserPolicyById(policyId: string) {
...
     contract.queryFilter(contract.filters.CreatedPolicy(null, userAccount, policyId)),
     contract.queryFilter(contract.filters.UpgradedPolicy(null, userAccount, policyId)),
+    contract.queryFilter(contract.filters.DowngradedPolicy(null, userAccount, policyId)),
   ]);
 
   if (!createdEvents.length) {
     return { ids: [], byId: {} };
   }
 
+  const stateChangedEvents = sortEvents([...upgradedEvents, ...downgradedEvents] as const);
```
Comment (Siegrift) [inline src/logic/policies/data.ts:137]: Nit: Do you need as const here?

File: src/utils/logs.test.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    expect(result).toEqual([
+      { testId: 'A', blockNumber: 0, logIndex: 0 },
+      { testId: 'B', blockNumber: 1, logIndex: 0 },
+      { testId: 'C', blockNumber: 2, logIndex: 1 },
+      { testId: 'D', blockNumber: 2, logIndex: 2 },
+      { testId: 'E', blockNumber: 2, logIndex: 3 },
+      { testId: 'F', blockNumber: 3, logIndex: 0 },
+      { testId: 'G', blockNumber: 4, logIndex: 0 },
+    ]);
+    expect(result === events).toBe(false); // Sanity check. The function should return a new array
```
Comment (Siegrift) [inline src/utils/logs.test.ts:26]: Nit:
```suggestion
    expect(result === events).toBe(false); // sortEvents is immutable
```

might be more descriprive.

**Review by Siegrift — APPROVED** [review]
👍 LGTM

File: src/logic/policies/data.ts (hunk 2 above)
Comment (mcoetzee) [reply src/logic/policies/data.ts:137]: Yeah this one was unexpected. Without `as const` TS incorrectly infers `stateChangedEvents` as `UpgradedPolicyEvent[]`. With `as const` TS correctly infers it as `UpgradedPolicyEvent | DowngradedPolicyEvent)[]`.

File: hardhat/hardhat.config.ts (hunk 1 above)
Comment (mcoetzee) [reply hardhat/hardhat.config.ts:134]: Yeah it accepts `BigNumberish` arguments here, to the `BigNumber.from(...)` is unnecessary

File: src/logic/policies/data.ts (hunk 2 above)
Comment (Siegrift) [reply src/logic/policies/data.ts:137]: Oh, I see. That seems one of the many bugs they have.

**Pull request merged**